virtualizarr/manifests/store.py

```diff
@@ -0,0 +1,367 @@
+        return dict_to_buffer(metadata, prototype=default_buffer_prototype())
+
+
+def parse_manifest_index(
+    key: str, chunk_key_encoding: str = "."
+) -> tuple[str, tuple[int, ...]]:
+    """
+    Splits `key` provided to a zarr store into the variable indicated
+    by the first part and the chunk index from the 3rd through last parts,
+    which can be used to index into the ndarrays containing paths, offsets,
+    and lengths in ManifestArrays.
+
+    Currently only works for 1d+ arrays with a tree depth of one from the
+    root Zarr group.
+
+    Parameters
+    ----------
+    key : str
+
+    Returns
+    -------
+    ManifestIndex
+    """
+    parts = key.split("/")
+    var = parts[0]
+    # Assume "c" is the second part
+    # TODO: Handle scalar array case with "c" holds the data
+    if chunk_key_encoding == "/":
+        indexes = tuple(int(ind) for ind in parts[2:])
+    else:
+        indexes = tuple(int(ind) for ind in parts[2].split(chunk_key_encoding))
+    return var, indexes
+
+
+def find_matching_store(stores: dict[str, Any], request_key: str) -> StoreRequest:
+    """
+    Find the matching store based on the store keys and the beginning of the URI strings,
+    to fetch data from the appropriately configured ObjectStore.
+
+    Parameters:
+    -----------
+    stores : dict
+        A dictionary with URI prefixes for different stores as keys
+    request_key : str
+        A string to match against the dictionary keys
+
+    Returns:
+    --------
+    StoreRequest
+    """
+    # Sort keys by length in descending order to ensure longer, more specific matches take precedence
+    sorted_keys = sorted(stores.keys(), key=len, reverse=True)
+
+    # Check each key to see if it's a prefix of the uri_string
+    for key in sorted_keys:
+        if request_key.startswith(key):
+            return StoreRequest(store_id=key, key=request_key[len(key) :])
+    # if no match is found, raise an error
+    raise ValueError(
+        f"Expected the one of stores.keys() to match the data prefix, got {stores.keys()} and {request_key}"
+    )
+
+
+class ManifestStore(Store):
+    """A Zarr store that uses obstore for fast read/write from AWS, GCP, Azure.
+
+    Parameters
+    ----------
+    stores : dict[prefix, obstore.store.ObjectStore]
+        A mapping of url prefixes to obstore Store instances set up with the proper credentials.
+
+    Warnings
+    --------
+    ManifestStore is experimental and subject to API changes without notice. Please
+    raise an issue with any comments/concerns about the store.
+
+    Notes
+    -----
+    Modified from https://github.com/zarr-developers/zarr-python/pull/1661
+    """
+
+    _manifest_group: ManifestGroup
+    _stores: StoreDict
+
+    def __eq__(self, value: object):
+        NotImplementedError
+
+    def __init__(
+        self,
+        manifest_group: ManifestGroup,
+        *,
+        stores: StoreDict,
+    ) -> None:
```

Comment:
A docstring here would be helpful to explain that I think this `stores` argument works by prefix?